### DISCUSSION TITLE
autolearn fuzzy when score above X

### DISCUSSION THREAD

**Discussion by maxir3** [rspamd/rspamd#4897] [General]
Hi,
I would like to have rspamd add mails to fuzzy storage with a specific flag automatically when the score is above X. In my mind this could be useful to catch mails with a very similar content to spam messages in the past which now come from a fresh ip which is not yet listed in the rbls I use.

As far as I got it sorted out I should set up a script referred to in learn_condition which should return true only when a specific score is reached for that message. But I don't know how to accomplish that. Maybe someone can give my some advice?

Or do I have to use a post-filter rule?

**Comment by fatalbanana** [comment]
The way to do it is with an `idempotent`  postfilter - that's the stage where score could be checked; `learn_condition` is for filtering learns, not initiating them.

The problem is that you will potentially be learning from mistakes & re-enforcing those via fuzzy.

~~~lua
-- /etc/rspamd/rspamd.local.lua

local FUZZY_WEIGHT = 1
local FUZZY_FLAG = 5
local MIN_SCORE = 5

rspamd_config.LEARN_FUZZY_BY_SCORE = {
  callback = function(task)
    local sc = task:get_metric_score()
    if sc[1] < MIN_SCORE then return end
    rspamd_plugins.fuzzy_check.learn(task, FUZZY_FLAG, FUZZY_WEIGHT)
  end,
  type = 'idempotent'
}
~~~

**Reply by cleicht** [reply]
Is it enough to create the file “rspamd.local.lua” and the code, or do you have to enter something else?
The file alone hasn't produced any results.
No local fuzzy data is created. Of course I have configured a local fuzzy, but I'm currently only learning it by hand.
What else do I have to do for automatic training?

**Reply by maxir3** [reply]
I got it working in the meantime with a slightly different approach including saving the raw message to a tmp file and calling rspamc as a system command... Thanks to fatalbanana I know learned a much easier solution. This is my (now tuned) version, which I can confirm does work. It will give an "error" message in the history tab of rspamd's webinterface on every learned message. On fail, it will just say "false", you can investigate by looking into /var/log/rspamd/rspamd.log:

  cat /var/log/rspamd/rspamd.log | grep -i fuzzy

Lua script:

```lua
 -- /etc/rspamd/rspamd.local.lua

rspamd_config.AUTOLEARNFUZZY = {
	callback = function(task)
		local res = task:get_metric_result()
		local subj = task:get_subject()
	
		if res.score > 30 then
			local command = rspamd_plugins.fuzzy_check.learn(task, 15, 5)
			rlogger.errx('fuzzy auto learn: %1 (%2)', command, subj)
		end
	end,
	type = 'postfilter',
	priority = 1,
	description = 'auto learn fuzzy storage',
}
```

You might want to adjust the min score (30) and the learned weight (5).

And of course you have to define the Fuzzy Flag (i use 15):

```lua
  -- /etc/rspamd/local.d/fuzzy_check.conf

rule "local" {
	# Fuzzy storage server list
	servers = "localhost:11335";
	# Additional mime types to store/check
	mime_types = ["*"];
	# Hash weight threshold for all maps
	max_score = 25.0;
	# Whether we can learn this storage
	read_only = false;
	# Ignore unknown flags
	skip_unknown = yes;
	# Hash generation algorithm
	algorithm = "mumhash";
	# Use direct hash for short texts
	short_text_direct_hash = true;

	# Map flags to symbols
	fuzzy_map = {
		LOCAL_FUZZY_MANUAL {
			# Local threshold
			max_score = 30.0;
			# Flag to match
			flag = 11;
		}
		LOCAL_FUZZY_USERLEARN {
			max_score = 20.0;
			flag = 12;
		}
		LOCAL_FUZZY_AUTOLEARN {
			# Local threshold
			max_score = 30.0;
			# Flag to match
			flag = 15;
		}
	}
}
```

And the symbols:

```lua
-- /etc/rspamd/local.d/fuzzy_group.conf

symbols = {
    "LOCAL_FUZZY_MANUAL" {
        weight = 20.0;
        description = "Manually added fuzzy hash";
    }
    "LOCAL_FUZZY_USERLEARN" {
        weight = 5.0;
        description = "User added fuzzy hash";
    }
     "LOCAL_FUZZY_AUTOLEARN" {
        weight = 15;
        description = "autolearned fuzzy hash";
    }
}
```

Check config und reload rspamd if everything is fine:

rspamadm configtest
systemctl reload rspamd.service

**Reply by cleicht** [reply]
Thanks, that works very well.
Does it make sense to create this for Ham too?

if res.score > 30 then
			local command = rspamd_plugins.fuzzy_check.learn(task, 15, 5)
			rlogger.errx('fuzzy auto learn: %1 (%2)', command, subj)
		end
if res.score < 0 then
			local command = rspamd_plugins.fuzzy_check.learn(task, 12, 5)
			rlogger.errx('fuzzy auto learn: %1 (%2)', command, subj)
		end

and the associated fuzzy_check.conf and fuzzy_group.conf settings
Christian

**Comment by fatalbanana** [comment]
> Does it make sense to create this for Ham too?

No, fuzzy is not a classifier like bayes.

Further, changing flag does not reset weight which is not likely expected.

**Comment by DiveSurfer** [comment]
hello, 
i'm in the process of creating a script that only calls the fuzzy learn for a certain value (>5 and <15). after searching the net i came across this article. unfortunately i must have a bug in the script, because fuzzy is trained for every email.
does anyone have an idea why this is the case? 

```lua
 -- /etc/rspamd/rspamd.local.lua

local FUZZY_WEIGHT = 5
local FUZZY_FLAG = 11
local MIN_SCORE = 15
local MAX_SCORE = 6

rspamd_config.LEARN_FUZZY_BY_SCORE = {
  callback = function(task)
    local sc = task:get_metric_score()
    if (sc[1] < MIN_SCORE and sc[1] > MAX_SCORE) then return end
    rspamd_plugins.fuzzy_check.learn(task, FUZZY_FLAG, FUZZY_WEIGHT)
  end,
  type = 'idempotent'
}
```
alternatively i had already tried to train fuzzy only for the action “add header” (which would be enough for me), which leads to the same result that every message gets a fuzzy hash
```lua
local action = task:get_metric_action('default')
if action == 'add header' then return end
```

**Reply by DiveSurfer** [reply]
I was able to find my mistake, the “return end” does not belong to the if.

**Comment by syntaxys** [comment]
I stumbled across this thread while doing some research and would like to show you how I solved it in my case:

```lua
-- /etc/rspamd/rspamd.local.lua
-- Include extra lua files from $LOCAL_CONFDIR/local.d/lua.d/*.lua

local local_conf = rspamd_paths['LOCAL_CONFDIR']
local f = io.popen("ls -1 " .. local_conf .. "/local.d/lua.d/*.lua")

if f then
  for mod in f:lines() do
    dofile(mod)
  end
end

```
The previous snippet is only needed if you manage your code in separate files. That's your fuzzy feeder:
```lua
-- /etc/rspamd/local.d/lua.d/local_fuzzy_autolearn.lua
-- Local Fuzzy autolearn
local FUZZY_WEIGHT_SPAM = 12
local FUZZY_FLAG_SPAM = 11
local MAX_SCORE_SPAM = 30
local MIN_SCORE_SPAM = 12

local FUZZY_WEIGHT_PROB = 4
local FUZZY_FLAG_PROB = 12
local MAX_SCORE_PROB = 11.99
local MIN_SCORE_PROB = 8

local FUZZY_WEIGHT_HAM = -2
local FUZZY_FLAG_HAM = 13
local MAX_SCORE_HAM = 4
local MIN_SCORE_HAM = -2

local function isMilter (tab, val)
    for index, value in ipairs(tab) do
        if value == val then
            return true
        end
    end
    return false
end

rspamd_config.LEARN_FUZZY_LOCAL = {
  callback = function(task)
    local sc = task:get_metric_score()
    local fl = task:get_flags()

    if isMilter(fl, 'milter') then
        if (sc[1] <= MAX_SCORE_SPAM and sc[1] >= MIN_SCORE_SPAM)
            then rspamd_plugins.fuzzy_check.learn(task, FUZZY_FLAG_SPAM, FUZZY_WEIGHT_SPAM)
        elseif (sc[1] <= MAX_SCORE_PROB and sc[1] >= MIN_SCORE_PROB)
            then rspamd_plugins.fuzzy_check.learn(task, FUZZY_FLAG_PROB, FUZZY_WEIGHT_PROB)
        elseif (sc[1] <= MAX_SCORE_HAM and sc[1] >= MIN_SCORE_HAM)
            then rspamd_plugins.fuzzy_check.learn(task, FUZZY_FLAG_HAM, FUZZY_WEIGHT_HAM)
        else return end
    else return end
  end,
  type = 'idempotent'
  }
```
**EDIT:**
I have noticed that with the first version _auto_learn_ is also triggered if you only scan a mail in the web frontend. This is not intended, so I added a condition that checks whether a task is initiated through the milter channel.

And further more a hint:
If you train your HAM flag with negative weight values, your symbol should have a positive weight value to get negative scores 😉
```
# /etc/rspamd/local.d/fuzzy_group.conf

symbols = {
    "LOCAL_FUZZY_UNKNOWN" {
        weight = 5.0;
        description = "Generic local fuzzy hash match";
    }
    "LOCAL_FUZZY_SPAM" {
        weight = 12.0;
        description = "Blacklisted local fuzzy hash";
    }
    "LOCAL_FUZZY_PROB" {
        weight = 5.0;
        description = "Probable junk local fuzzy hash";
    }
    "LOCAL_FUZZY_HAM" {
        weight = 3.0;
        description = "Whitelisted local fuzzy hash";
    }
}
```
This results in e. g.:
`LOCAL_FUZZY_HAM (-2.414506) [13:bb0f27ae86:0.84:txt]` or
`LOCAL_FUZZY_SPAM (12.514309) [11:210246438e:1.00:txt,11:4bb5e84814:1.00:img]`
